### DISCUSSION TITLE
Build ASMC

### DISCUSSION THREAD

**Discussion by jklwn4** [nidud/asmc#5] [Q&A] [answered]
Trying to build ASMC in VC 22 with asmc.vcxproj fails because of path problems. I´m not an expert for VC, so maybe i´m doing something wrong. I expected opening the project would be enough for being able to rebuilt ASMC. VC doesn´t find asmc.props because it expects it to be in "C:\bin\asmc.props". I have no idea what to is wrong.

**Comment by nidud** [comment]
> maybe i´m doing something wrong.

No, you definitely on the right track here ;-)

The short answer is that you should launch VS from the shell (`./asmc-2.34.cmd` or `./bin/dz.exe`) as this holds the local environment for Asmc.

Assuming from the question this is the first time you test the debugging features in Asmc here are a few things to consider and prepare for.

The tool chain (like [Mingw](https://www.mingw-w64.org/) and others) are made for local installations which means you could have multiple versions of Asmc installed on different drives and directories, hence adding a global environment (like LIB, INCLUDE, PATH) for Asmc are not recommended.

The package should be installed using the asmc-x.xx.cmd file in the root directory, and a shortcut added to the desktop for this directory (to ./bin/dz.exe) named Asmc-x.xx with admin rights. It's also recommended that you exclude this directory from live AV scans as this slows down the build process, especially on large projects and libraries.

Take some time to setup DZ properly. Set the right font (and) size, test the Alt-F9 and Alt-Enter, and finally hit Alt-T and enter. The default PATH setting should be:
```
[Path]
0=%DZ%
1=%PATH%
```

Here you should add the path to the VS tools (if not already in the PATH):
```
[Path]
0=%DZ%
1=<VS2022>\VC\Tools\MSVC\14.36.32532\bin\Hostx64\x64
2=<VS2022>\VC\Tools\MSVC\14.36.32532\bin\Hostx64\x86
3=%PATH%
```

Exit and restart the shell for this to take effect.

Now to the last bit (debated before) regards the import libraries: Navigate to the `./lib` directory and hit enter on the` VS2019.mak` file to build the imports using VS tools.

Note that it is a great convenience to have two screens when debugging. One for the source and one for registers, call stack and watches. 

To test if this works try a simple project first, like `../test/libc/stdio/fopen`. Hit enter on the project file, open the source and put a break point at the first line (F9) and hit F5. Then use F11 to trace into the source and see how that goes.

Debugging Asmc needs source input so you need to add a `test.asm` in the current directory and modify the command line setting in the project config.

**Reply by jklwn4** [reply]
I´m not very familiar with VC and not familar with dz.exe at all. I have my own IDE, which i can adapt to my wishes and needs. Good to know how i can debug ASMC with VC. 
My first goal was to be able to re-build ASMC for (Windows version 32 and 64 bit) itself, a simple batch file would be extremly convenient here. I would like to apply a few small changes and additions, at least i would like to try this. For several reasons i prefer the D:\ drive, so it´s D:\asmc for me.

**Comment by jklwn4** [comment]
Trying to follow your above steps, i can create a 64 bit debug version of ASMC in VC

Using dz.exe it fails (makefile in \source\asmc), it assembles, but then i get this:

        ..\..\bin\linkw @C:\Users\Juergen\AppData\Local\Temp\$8E8C11$.$$$                                                                                                                                       del *.obj                                                                                                                                                                                       Warning! W1008: cannot open uuid.lib : No such file or directory                                                                                                                                        Error! E2028: __flsbuf is an undefined reference                                                                                                                                                        Error! E2028: ___pioinfo is an undefined reference                                                                                                                                                      Error! E2028: __nfile is an undefined reference                                                                                                                                                         Error! E2028: __alloc_osfhnd is an undefined reference                                                                                                                                                  file libc.lib(fwrite.obj): undefined symbol __flsbuf                                                                                                                                                    file libc.lib(rewind.obj): undefined symbol ___pioinfo                                                                                                                                                  file libc.lib(ftell.obj): undefined symbol ___pioinfo                                                                                                                                                   file libc.lib(_write.obj): undefined symbol ___pioinfo                                                                                                                                                  file libc.lib(_close.obj): undefined symbol ___pioinfo                                                                                                                                                  file libc.lib(_close.obj): undefined symbol __nfile                                                                                                                                                     file libc.lib(_read.obj): undefined symbol ___pioinfo
file libc.lib(_filbuf.obj): undefined symbol ___pioinfo
file libc.lib(_sopen.obj): undefined symbol __alloc_osfhnd
file libc.lib(_get_osfhandle.obj): undefined symbol ___pioinfo
file libc.lib(_get_osfhandle.obj): undefined symbol __nfile

**Comment by nidud** [comment]
> Trying to follow your above steps, i can create a 64 bit debug version of ASMC in VC

Good, so now you can try the debugging features (greatly) needed to add any changes.

> Using dz.exe it fails (makefile in \source\asmc), it assembles, but then i get this:

The makefile was out of date so I just uploaded a new one. This defaults to 64-bit so additional arguments to make is `x86=1` and `ASMC64=1`. 

The command for building Asmc in Windows should be something like this (make sure asmc is not in this directory):
```
git clone https://www.github.com/nidud/asmc.git
cd asmc
asmc-2.34
cd source\asmc
make
```

Commands for Linux:
```
git clone https://www.github.com/nidud/asmc.git
cd asmc/source/asmc/gcc
make
make install
```

**Comment by jklwn4** [comment]
Thanks - now i can build ASMC for 64 bit but for 32 bit it still fails with the same errors as above.

**Comment by nidud** [comment, marked as answer]
Added a new [update](https://github.com/nidud/asmc/commit/a32d99a43786533e3700e387f7ae026067103277) with some 32-bit updates.

**Reply by jklwn4** [reply]
Many thanks - now it works! I can build both a 32 bit version and a 64 bit version and both assemble my testsuite flawlessly, great!
